doc/source/signals.rst

```diff
@@ -0,0 +1,183 @@
+Signal API (Experimental)
+=========================
+
+This experimental API allows tasks and actors to generate signals which can
+be received by other tasks and actors. In addition, task failures and actor
+method failures generate error signals. The error signals enable applications
+to detect failures and potentially recover from failures.
+
+.. autofunction:: ray.experimental.signal.send
+
+Here is a simple example of a task sending a user-defined signal.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal 
+
+  # Define an application level signal.
+  class UserSignal(signal.Signal):
+      def __init__(self, value):
+            self.value = value
+
+      def get_value(self):
+            return self.value
+
+  # Define a remote function that sends a user-defined signal.
+  @ray.remote
+  def send_signal(value):
+      signal.send(UserSignal(value))
+
+.. autofunction:: ray.experimental.signal.receive
+
+Here is a simple example of how to receive signals from an actor or task identified
+by ``a``. Note that an actor is identified by its handle, and a task by one of its
+object ID return values.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal
+
+  # This returns a possibly empty list of all signals that have been sent by 'a'
+  # since the last invocation of signal.receive from within this process. If 'a'
+  # did not send any signals, then this will wait for up to 10 seconds to receive
+  # a signal from 'a'.
+  signal_list = signal.receive([a], timeout=10)
+
+.. autofunction:: ray.experimental.signal.forget
+
+Below is a simple example, showing how to ignore all signals sent so
+far by ``a`` and only wait for new signals from ``a``.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal
+
+  signal.forget([a])
+
+  signal_list = signal.receive([a], timeout=10)
+
+
+.. autofunction:: ray.experimental.signal.reset
+
+
+Example: sending a user signal
+------------------------------
+
+The code below show a simple example in which a task, called ``send_signal()``
+sends a user signal and the driver gets it by invoking ``signal.receive()``.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal
+
+  # Define a user signal.
+  class UserSignal(signal.Signal):
+      def __init__(self, value):
+            self.value = value
+
+      def get_value(self):
+            return self.value
+
+  @ray.remote
+  def send_signal(value):
+      signal.send(UserSignal(value))
+      return
+
+  signal_value = 'simple signal'
+  object_id = send_signal.remote(signal_value)
+  # Wait up to 10sec to receive a signal from the task. Note the task is
+  # identified by the object_id it returns.
+  result_list = signal.receive([object_id], timeout=10)
+  # Print signal values. This should print "simple_signal".
+  # Note that result_list[0] is the signal we expect from the task.
+  # The signal is a tuple where the first element is the first object ID 
```

Comment:
Which object ID is it when the task has multiple return values?

Reply:
You can use any of the object IDs returned by a task.  

If you put in the receive list multiple object IDs returned by the same task, you get the same signals for each of these object IDs.

Reply:
I'm asking about the value *returned* by `signal.receive`. Not the argument to `signal.receive`.